docs_src/design/adr/security/0019-spiffe.md

```diff
@@ -0,0 +1,158 @@
+   to verify the client,
+   extracts the service key,
+   and issues an appropriate Vault service token.
+1. The EdgeX microservice accesses Vault as usual.
+
+## Consequences
+
+This proposal will require addition of several new, optional, EdgeX microservices:
+
+* `security-spiffe-token-provider`, running on the main node
+* `spiffe-agent`, running on the main node and each remote node
+* `spiffe-server`, running on the main node
+* `spiffe-config`, a one-shot service running on the main node
+
+Minor changes will be needed to `security-secretstore-setup` to preserve
+the token-creating-token used by `security-file-token-provider` so that
+it can be used by `security-spiffe-token-provider`.
+
+The startup flow of the framework will be adjusted as follows:
+
+* Bootstrap service (original)
+  * `spiffe-server`
+    * `spiffe-config` (can be combined with `spifee-server`)
+      * `spiffe-agent`
+  * Vault service (original)
+    * Secret store setup service (original)
+      * `security-spiffe-token-provider`
+    * Consul (original)
+    * Postgres (original)
+
+There is no direct dependency between `spiffe-server`
+and any other microservice.
+`security-spiffe-token-provider` requires an SVID
+from `spiffe-agent` and a Vault admin token.
+
+None of these new services will be proxied via the API gateway.
+
+A new configuration file will be added to `spiffe-config`
+to configure the workload authorization parameters.
+This configuration is dependent upon whether one is running
+in a Snap configuration or a Docker configuration.
+For Snap configurations, new services with be authenticated
+using the Unix file path of the executable.
+For Docker configurations, the new services will be authenticated
+using the Docker service name (coded into the running container as a label).
+The configuration file will be pre-loaded with known EdgeX services
+as they are today.
+Additional services can be loaded into the authorization database
+by injecting additional configuration files,
+or on-demand by executing a command line utility.
+It is no longer necessary to restart the EdgeX framework
+to authenticate a new service to the EdgeX secret store.
+
+In the future,
+this mechanism may become the default secret store distribution mechanism,
+as it eliminates several secrets volumes used to share secrets between
+`security-secretstore-setup` and various EdgeX microservices.
+
+The EdgeX automation will only configure the SPIFEE agent on the main node.
+Additional nodes can be manually added by the operator
+by obtaining a join token from the main node
+and using it to bootstrap a remote node.
+
+This solution only works on Linux platforms,
+as the SPIRE server and SPIRE workload attester have not been to Windows.
+
+SPIFFE/SPIRE has native support for Kubernetes
+and can distribute the trust bundle via a Kubernetes
+ConfigMap to more easily enable distributed scenarios,
+removing a major roadblock to usage of EdgeX
+in a Kubernetes environment.
+
+
```

Comment:
How dose this approach impact the ability to specify known secrets be added to service's Secret Store.

Reply:
> How dose this approach impact the ability to specify known secrets be added to service's Secret Store.

There is no support for it.

If this is OK there is nothing else to be done.

if this is NOT OK then I am open to proposals to handle propagation of known secrets. Perhaps the client can request propagation of known secrets when requesting the secret store token?

Reply:
I think its important since we've moved to Secure Message bus by default...  Also what about Consul ACL roles?

Reply:
Ok.  Will add ability to request known secrets to the API.